efi/image_rules_defs.go

```diff
@@ -0,0 +1,202 @@
+			),
+			newShimLoadHandler,
+		),
+		// Ubuntu signed 2 shim 15's based on commit a4a1fbe with the required patches
+		// backported. These were the initial UC20 shims
+		withImageRule(
+			"Ubuntu shim 15 with required patches",
+			imageMatchesAll(
+				imageSectionExists(".vendor_cert"),
+				shimVersionIs("==", "15"),
+				imageMatchesAny(
+					// shim-signed_1.40.4+15+1552672080.a4a1fbe-0ubuntu2_amd64.deb
+					imageDigestMatches(crypto.SHA256, []byte{
+						0x2e, 0xa4, 0xcb, 0x6a, 0x1f, 0x1e, 0xb1, 0xd3,
+						0xdc, 0xe8, 0x2d, 0x54, 0xfd, 0xe2, 0x6d, 0xed,
+						0x24, 0x3b, 0xa3, 0xe1, 0x8d, 0xe7, 0xc6, 0xd2,
+						0x11, 0x90, 0x2a, 0x59, 0x4f, 0xe5, 0x67, 0x88,
+					}),
+					// shim-signed_1.41+15+1552672080.a4a1fbe-0ubuntu1_amd64.deb
+					imageDigestMatches(crypto.SHA256, []byte{
+						0xe0, 0x60, 0xda, 0x09, 0x56, 0x1a, 0xe0, 0x0d,
+						0xcf, 0xb1, 0x76, 0x9d, 0x6e, 0x8e, 0x84, 0x68,
+						0x68, 0xa1, 0xe9, 0x9a, 0x54, 0xb1, 0x4a, 0xa5,
+						0xd0, 0x68, 0x9f, 0x28, 0x40, 0xce, 0xc6, 0xdf,
+					}),
+				),
+			),
+			newShimLoadHandlerConstructor().WithVersion(mustParseShimVersion("15.2")).New,
+		),
+		// Pre SBAT shims - unsupported. These will cause an error from
+		// newShimLoadHandler rather than allowing this to fallback to the
+		// null handler.
+		withImageRule(
+			"unsupported shim",
+			imageSectionExists(".vendor_cert"),
+			newShimLoadHandler,
+		),
+		// Ubuntu Grub with chainloader patch
+		withImageRule(
+			"Ubuntu grub",
+			imageMatchesAny(
+				imageMatchesAll(
+					sbatSectionExists,
+					sbatComponentExists("grub.ubuntu"),
+				),
+				imageMatchesAll(
+					imageSectionExists(".mods"),
+					imageSignedByOrganization("Canonical Ltd."),
+				),
+			),
+			newGrubLoadHandlerConstructor(grubChainloaderUsesShimProtocol).New,
+		),
+		withImageRule(
+			"grub",
+			imageMatchesAny(
+				imageMatchesAll(
+					sbatSectionExists,
+					sbatComponentExists("grub"),
+				),
+				imageSectionExists(".mods"),
+			),
+			newGrubLoadHandler,
+		),
+		withImageRule(
+			"Ubuntu Core UKI",
+			imageMatchesAny(
+				imageMatchesAll(
+					sbatSectionExists,
+					sbatComponentExists("systemd.ubuntu"),
+				),
+				imageMatchesAll(
+					// TODO: Add another primitive here to check the contents of the
+					// .sdmagic section as an alternative to checking for the existence
+					// of the .linux and .initrd sections.
+					imageSectionExists(".linux"),
+					imageSectionExists(".initrd"),
+					imageSignedByOrganization("Canonical Ltd."),
+				),
+			),
+			newUbuntuCoreUKILoadHandler,
+		),
+	)
+}
+
+func makeFallbackImageRules() *imageRules {
+	return newImageRules(
+		"Fallback",
+		// Shim
+		newImageRule(
+			"shim",
+			imageSectionExists(".vendor_cert"),
+			newShimLoadHandler,
+		),
+		// Grub
+		newImageRule(
+			"grub",
+			imageSectionExists(".mods"),
+			newGrubLoadHandler,
+		),
+		// TODO: add rules for Ubuntu Core UKI and Ubuntu grub that are not part of
+		// the MS UEFI CA?
+		//
```

Comment:
what's the state of things, do we need to fix this TODO now?

Reply:
We don't, but I'll probably add a section to our UKI to allow it to be identified more easily